### DISCUSSION TITLE
Problems with Normalize node

### DISCUSSION THREAD

**Discussion by hgmarques** [bonsai-rx/bonsai#1397] [Q&A]
Hi guys,

I've noticed that the Normalize node not always converts the image pixels to values in the range of [0,1] (or at least the IplImageVisualizer on the Normalize node shows values outside of the [0,1] range, image attached). I don't really understand why since the source code seems to be fine. 

I attached a simple Bonsai example to reproduce my observations.

Any help would be welcome. Thanks you in advance.

Cheers,
hugo

![NormalizeBug](https://github.com/bonsai-rx/bonsai/assets/4449346/d8225115-9ed8-425a-80a3-61a5653ef199)
[NormalizeBug.zip](https://github.com/bonsai-rx/bonsai/files/11627667/NormalizeBug.zip)

**Comment by bruno-f-cruz** [comment]
It looks funky. Looking at the function bonsai implements:

https://github.com/bonsai-rx/bonsai/blob/404b092b501fbd9a0db5ccce76bf695623e6ba7c/Bonsai.Vision/Normalize.cs#LL34C17-L34C32
```csharp
                var range = max - min;
                var scale = range > 0 ? 1.0 / range : 0;
                var shift = range > 0 ? -min : 0;
                CV.ConvertScale(input, output, scale, shift);
```

`shift` is always in raw units.
However, the [opencv function](https://docs.opencv.org/3.4/d2/df8/group__core__c.html#ga3a6c66666ac8c30f639c85dd18df9f67) follows the canonical algebraic order of operation and applies the scaling first and the shift second.

![image](https://github.com/bonsai-rx/bonsai/assets/7049351/fb9a89fd-b230-4538-875d-48b4add6e594)


It seems the easiest way to correct this would be to normalize the shift by the range:
```csharp
var shift = range > 0 ? -min*scale : 0;
```

Here's a temporary fix that you can place in your extensions folder.

```csharp
﻿using OpenCV.Net;
using System;
using System.ComponentModel;
using System.Linq;
using System.Reactive.Linq;

namespace Bonsai.Vision
{
    /// <summary>
    /// Represents an operator that normalizes the range of values for each image
    /// in the sequence to be between zero and one.
    /// </summary>
    [Description("Normalizes the range of values for each image in the sequence to be between zero and one.")]
    public class NormalizeFix : Transform<IplImage, IplImage>
    {
        /// <summary>
        /// Normalizes the range of values for each image in an observable sequence
        /// to be between zero and one.
        /// </summary>
        /// <param name="source">The sequence of images to normalize.</param>
        /// <returns>
        /// The sequence of normalized images.
        /// </returns>
        public override IObservable<IplImage> Process(IObservable<IplImage> source)
        {
            return source.Select(input =>
            {
                var output = new IplImage(input.Size, IplDepth.F32, input.Channels);
                double min;
                double max;
                Point minLoc;
                Point maxLoc;
                CV.MinMaxLoc(input, out min, out max, out minLoc, out maxLoc);
                var range = max - min;
                var scale = range > 0 ? 1.0 / range : 0;
                var shift = range > 0 ? -min*scale : 0;
                CV.ConvertScale(input, output, scale, shift);
                return output;
            });
        }
    }
}
```
